ophyd/pseudopos.py

```diff
@@ -914,3 +912,131 @@ def set(self, position, **kwargs):
+    values are with 1% of the 'location'.   
+     
+    Parameters
+    ----------
+    locations : dictionary.
+        A keyword:Value dictionary that lists all of the predefined positions (keyword) and a 
+        dictionary of axis-value pairs to be set in this location in the form: 
+        {location1:[axis1:value1,axis2:value2,...], 
+            location2:[axis1:value1,axis2:value2,...],.....}.
+            NOTE: 
+            1. Not all axes need to have a specifed value for each device location, only 
+            those with a specifed value are moved/checked for a given location.
+            2. All axes specifed in this dictionary must be specifed as components in 'components'.
+    neighbours : Dictionary, optional.
+        A keyword:value dictionary where each keyword is a location defined in 'locations' and 
+        each value is a list of 'neighbours' for that location. Motion can only occur between 
+        neighbours, for non-neighbours a path through various locations will be used, if 
+        it is found using self.find_path. Optionally if the list contains 'All' this indicates 
+        that evey location is accesible from this one. If no neighbours are defined for this 
+        location then it is assumed that no direct motion is allowed from this location.
+    regions : Dictionary, optional.
+        The optional keyword:value dictionary that has location keywords and 'region' values 
+        'region' is a dictionary that has axis_name keywords and [min_val, max_val] values 
+        denoting the range of values for this location and axis. This dictionary has the form: 
+            {location1:{'axis1_name':[axis1_min_val,axis1_max_val],
+                                                   'axis2_name':[axis2_min_val,axis2_max_val],...}, 
+            location2:{'axis1_name':[axis1_min_val,axis1_max_val],
+                                                   'axis2_name':[axis2_min_val,axis2_max_val],...},
+                                          ....}.
+            NOTE: Not all locations in the 'locations' require an entry in this dictionary and not 
+            all axes defined with a value in 'locations', for a given location, must have a 
+            range in this dictionary for the given location. For any axis that a 'range' is not 
+            provided a default range of =/- 1% is used.
+
+
+    NOTES ON PREDEFINED MOTION WITH NEIGHBOURS:
+    1. To ensure the motion to a predefined location always occurs when using neighbours to define
+        motion 'paths' it is best to ensure that the device is always in a 'location' by making sure
+        that motion can not move the device outside of all 'locations'.
+    2. To ensure that motion occurs always via a path then each 'point' in the path should be a 
+        location, and it should only have the neighbours that are before or after it in the required
+        path.
+
+    '''
+    def __init__(self, *,locations={}, neighbours={}, regions={}):
```

Comment:
Should this just take in a `DiGraph`? That seems maximally flexible.

Reply:
I am very reluctant to have it take in a DiGraph, because to non-coding background BL staff that is a confusing way to specify things. I tried at SIX and both BL staff there where completely confused with the structure of DiGraph and how to implement one. This set of dictionaries, they claim, was much easier for them to understand and modify when necessary. When discussing it with Eli and Klaus at ISS they also much preferred this system, as they found it easy to understand (although they would like an extra abstraction so that it takes in an excel file).

Of course I could write an 'external' function that takes in these dictionaries and returns a DiGraph, which is what happens internally at present. A second function could be written to read in an excel file and return the DiGraph to give options for how to specify the interactions.